[ANGLE] Fix Metal API deprecation warnings

djg opened the pull request:
#### 79b1ae56836e96f11fd466cbadbca4b3abd013ff
<pre>
[ANGLE] Fix Metal API deprecation warnings
<a href="https://bugs.webkit.org/show_bug.cgi?id=248457">https://bugs.webkit.org/show_bug.cgi?id=248457</a>
rdar://102750943

Reviewed by Kimmo Kinnunen.

macOS 13/iphoneOS 16.0 deprecated Metal APIs, causing warnings when building
ANGLE. Fix warnings associated with using deprecated features by using
preprocessor checks to not use those APIs when the targeted minimum version is
the same as, or later than, macOS 13/iphoneOS 16.0

* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_command_buffer.mm:
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_state_cache.mm:
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_utils.mm:
(rx::mtl::SupportsMacGPUFamily):
(rx::mtl::ComputeTotalSizeUsedForMTLRenderPipelineDescriptor):

Canonical link: <a href="https://commits.webkit.org/257184@main">https://commits.webkit.org/257184@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9a93c8b92bd483a7d85830d6b1534c0e3c6cb9f8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98096 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/7310 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/31245 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107557 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167821 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/7799 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/36073 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/90662 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/104165 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103748 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/5864 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84692 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/32984 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/87733 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89448 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/75919 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1283 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20880 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1246 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/22383 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6108 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44834 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2465 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2548 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41791 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_command_buffer.mm:
```diff
@@ -362,22 +362,34 @@ inline void SetVisibilityResultModeCmd(id<MTLRenderCommandEncoder> encoder,
     mtl::RenderStages stages = stream->fetch<mtl::RenderStages>();
     ANGLE_UNUSED_VARIABLE(stages);
-#if defined(__IPHONE_13_0) || defined(__MAC_10_15)
+#if defined(ANGLE_MTL_USE_RESOURCE_USAGE_STAGES_AVAILABLE) && !defined(ANGLE_MTL_USE_RESOURCE_USAGE_DEPRECATED)
```
I think the compile error shows that this one is a bit too complicated ifdefing..
Maybe have two arms and don't try to share that much code?

djg replied:
I was trying to be too clever.

kkinnunen-apple commented on Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_command_buffer.mm:
```diff
@@ -362,22 +362,34 @@ inline void SetVisibilityResultModeCmd(id<MTLRenderCommandEncoder> encoder,
     {
+#endif
         [encoder useResource:resource usage:usage stages:stages];
+#if !defined(ANGLE_MTL_USE_RESOURCE_USAGE_DEPRECATED)
```
I think the compile error is that you don't have the same condition here.
Still, maybe it's simpler to have
```
#if ...
  ANGLE_UNUSED_VARIABLE(stages);
  if (...)
  {
  }
#else
   [encoder useResource:resource usage:usage stages:stages];
#endif
```

kkinnunen-apple commented on Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_utils.mm:
```diff
@@ -1374,10 +1383,11 @@ bool SupportsMacGPUFamily(id<MTLDevice> device, uint8_t macFamily)
     // If device doesn't support [MTLDevice supportsFamily:], then use
     // [MTLDevice supportsFeatureSet:].
-#    if TARGET_OS_MACCATALYST
+#    if TARGET_OS_MACCATALYST || ANGLE_MTL_FEATURE_SET_DEPRECATED
```
Remember to update the #endif // comment for this :)

kkinnunen-apple reviewed (approved):
Yes, should upstream

djg committed:
[ANGLE] Fix Metal API deprecation warnings
https://bugs.webkit.org/show_bug.cgi?id=248457
rdar://102750943

Reviewed by Kimmo Kinnunen.

macOS 13/iphoneOS 16.0 deprecated Metal APIs, causing warnings when building
ANGLE. Fix warnings associated with using deprecated features by using
preprocessor checks to not use those APIs when the targeted minimum version is
the same as, or later than, macOS 13/iphoneOS 16.0

* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_command_buffer.mm:
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_state_cache.mm:
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_utils.mm:
(rx::mtl::SupportsMacGPUFamily):
(rx::mtl::ComputeTotalSizeUsedForMTLRenderPipelineDescriptor):

Canonical link: https://commits.webkit.org/257184@main
